crates/next-api/src/global_information.rs

```diff
@@ -1,26 +1,118 @@
         module_id_map: build_module_id_map(project).await?,
     };
     Ok(Vc::cell(Some(global_information)))
 }
 
+async fn process_module(
+    module: Vc<Box<dyn Module>>,
+    id_map: &mut HashMap<AssetIdent, ModuleId>,
+    used_ids: &mut HashSet<u64>,
+) -> Result<()> {
+    let ident = module.ident();
+
+    let hash = hash_xxh3_hash64(ident.to_string().await?);
+    let mut masked_hash = hash & 0xF;
+    let mut mask = 0xF;
+    while used_ids.contains(&masked_hash) {
+        if mask == 0xFFFFFFFFFFFFFFFF {
+            return Err(anyhow::anyhow!("This is a... 64-bit hash collision?"));
+        }
+        mask = (mask << 4) | 0xF;
+        masked_hash = hash & mask;
+    }
+
+    let hashed_module_id = ModuleId::String(masked_hash.to_string().into());
+
+    dbg!(
+        "process_module",
+        ident.await?.clone_value(),
+        &hashed_module_id,
+        mask
+    );
+
+    id_map.insert(ident.await?.clone_value(), hashed_module_id);
+    used_ids.insert(masked_hash);
+
+    Ok(())
+}
+
 async fn build_module_id_map(project: Vc<Project>) -> Result<HashMap<AssetIdent, ModuleId>> {
-    Ok(HashMap::new())
+    let mut module_id_map: HashMap<AssetIdent, ModuleId> = HashMap::new();
+    let mut used_ids: HashSet<u64> = HashSet::new();
+
+    let entrypoints = project.entrypoints().await?;
+    let mut root_modules: Vec<Vc<Box<dyn Module>>> = Vec::new();
+    root_modules.extend(entrypoints.pages_error_endpoint.root_modules().await?);
+    root_modules.extend(entrypoints.pages_app_endpoint.root_modules().await?);
+    root_modules.extend(entrypoints.pages_document_endpoint.root_modules().await?);
+
+    for (_, route) in entrypoints.routes.iter() {
+        match route {
+            Route::Page {
+                html_endpoint,
+                data_endpoint: _,
+            } => {
+                root_modules.extend(html_endpoint.root_modules().await?);
+            }
+            Route::PageApi { ref endpoint } => {
+                root_modules.extend(endpoint.root_modules().await?);
+            }
+            Route::AppPage(routes) => {
+                for route in routes {
+                    root_modules.extend(route.html_endpoint.root_modules().await?);
+                }
+            }
+            Route::AppRoute {
+                original_name: _,
+                endpoint,
+            } => {
+                root_modules.extend(endpoint.root_modules().await?);
+            }
+            Route::Conflict => {
+                tracing::info!("WARN: conflict");
+            }
+        }
+    }
+
+    let modules_iter = AdjacencyMap::new()
+        .skip_duplicates()
+        .visit(root_modules.iter().copied(), get_referenced_modules)
+        .await
+        .completed()?
+        .into_inner()
+        .into_reverse_topological();
+
+    for module in modules_iter {
+        process_module(module, &mut module_id_map, &mut used_ids).await?;
+    }
+
+    Ok(module_id_map)
 }
```

Comment:
This will visit the whole application at once. That might lead to problems.

Could we extract global information per page so we only process a page at a time and merge these data to the whole app global information?

Reply:
What kind of problems do you forsee? To me it seems much faster to visit a single large graph, as when we short circuit traversal on seen nodes we benefit from the previous page’s traversal and can stop before huge subgraphs are visited.